### PR TITLE
logistic regression burden tests

### PR DESCRIPTION
**Pull request by jbloom22**

### CODE REVIEW

File: src/main/scala/is/hail/methods/LogisticRegression.scala [hunk 1]
```diff
@@ -14,17 +14,16 @@ object LogisticRegression {
     require(vds.wasSplit)
 
     def tests = Map("wald" -> WaldTest, "lrt" -> LikelihoodRatioTest, "score" -> ScoreTest, "firth" -> FirthTest)
```
Comment (danking) [inline src/main/scala/is/hail/methods/LogisticRegression.scala]: This is a `val` and it shouldn't be inside the `apply` method, it should be a `private` `val` on the `LogisticRegression` object.

File: src/main/scala/is/hail/methods/LogisticRegression.scala [hunk 2]
```diff
@@ -14,17 +14,16 @@ object LogisticRegression {
...
+    if (!tests.isDefinedAt(test))
+      fatal(s"Supported tests are ${ tests.keys.mkString(", ") }, got: $test")
 
     val logRegTest = tests(test)
```
Comment (danking) [inline src/main/scala/is/hail/methods/LogisticRegression.scala]: I'm inclined to use this instead:

```scala
val logRegTest = tests.get(test).getOrElse {
    fatal(s"Supported tests are ${ tests.keys.mkString(", ") }, got: $test")
  }
```

it avoids the unsafe `apply` method.

File: src/main/scala/is/hail/methods/LogisticRegressionBurden.scala [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  def apply(vds: VariantDataset,
+    keyName: String,
+    variantKeys: String,
+    singleKey: Boolean,
+    aggExpr: String,
+    test: String,
+    yExpr: String,
+    covExpr: Array[String]): (KeyTable, KeyTable) = {
+
+    def tests = Map("wald" -> WaldTest, "lrt" -> LikelihoodRatioTest, "score" -> ScoreTest, "firth" -> FirthTest)
```
Comment (danking) [inline src/main/scala/is/hail/methods/LogisticRegressionBurden.scala]: This is repeated from above and is still not a `def` nor should it be local to this method. 

`LogisticRegressionTest` should have a companion object that has a map of all the tests in one place.

File: src/main/scala/is/hail/methods/LogisticRegressionBurden.scala [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    val d = n - k - 1
+
+    if (d < 1)
+      fatal(s"$n samples and $k ${ plural(k, "covariate") } including intercept implies $d degrees of freedom.")
+
+    val nullModel = new LogisticRegressionModel(cov, y)
+    val nullFit = nullModel.fit()
+
+    if (!nullFit.converged)
+      fatal("Failed to fit (unregulatized) logistic regression null model (covariates only): " + (
```
Comment (danking) [inline src/main/scala/is/hail/methods/LogisticRegressionBurden.scala]: unregularized ?

File: src/main/scala/is/hail/methods/LogisticRegressionBurden.scala [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+    val emptyStats = logRegTest.emptyStats
+
+    val logregRDD = sampleKT.rdd.mapPartitions({ it =>
+      val X = XBc.value.copy
+      it.map { keyedRow =>
+        val key = keyedRow.get(0)
+
+        if (RegressionUtils.setLastColumnToKeyedRow(X, keyedRow))
+          logRegTestBc.value.test(X, yBc.value, nullFitBc.value).toAnnotation(emptyStats, Some(key)).asInstanceOf[Row]
```
Comment (danking) [inline src/main/scala/is/hail/methods/LogisticRegressionBurden.scala]: Modifying `toAnnotation` to take a key argument that it appends to the `Row` violates the `TStruct` merge interface. Lift `logregSignature` up and make it into `val (logregSignature, merger) = ...` and use the merger to combine the key with the `Row`. Same for combining the key with the `emptyStats`.

Comment (danking) [inline src/main/scala/is/hail/methods/LogisticRegressionBurden.scala]: Actually, why is `emptyStats` passed in? It should be given to the `...Result` class by the `...Test` class. Having the consumer of the `LogisticRegressionTest` API supply the `emptyStats` argument to the `...Result` classes is prone to mistakes.

File: src/test/scala/is/hail/methods/LinearRegressionBurdenSuite.scala [hunk 6]
```diff
@@ -1,14 +1,31 @@
...
 import is.hail.SparkSuite
+import is.hail.TestUtils.interceptFatal
 import is.hail.expr.TDouble
 import is.hail.keytable.KeyTable
 import is.hail.utils._
 import is.hail.variant.VariantDataset
 import org.apache.spark.sql.Row
 import org.testng.annotations.Test
 
+object BurdenSuiteUtils {
```
Comment (danking) [inline src/test/scala/is/hail/methods/LinearRegressionBurdenSuite.scala]: Put this in a separate file, it's not specific to linear or logistic so it doesn't belong in either of the existing files..

File: src/test/scala/is/hail/methods/LinearRegressionBurdenSuite.scala [hunk 7]
```diff
@@ -235,4 +239,16 @@ class LinearRegressionBurdenSuite extends SparkSuite {
     assert(sampleMap8.size == 10)
     assert(sampleMap8.forall { case (key, value) => value.forall(_ % key.asInstanceOf[Double] == 0) })
   }
+
+  @Test def testFatals() {
```
Comment (danking) [inline src/test/scala/is/hail/methods/LinearRegressionBurdenSuite.scala]: All the changes to linreg, including these tests, should be a separate PR.

File: src/main/scala/is/hail/stats/RegressionUtils.scala [hunk 8]
```diff
@@ -335,7 +373,7 @@ object RegressionUtils {
   }
 
   //keyedRow consists of row key followed by numeric data (passed with key to avoid copying, key is ignored here)
-  def denseStats(keyedRow: Row, y: DenseVector[Double]): Option[(DenseVector[Double], Double, Double)] = {
+  def statsKeyedRow(keyedRow: Row, y: DenseVector[Double]): Option[(DenseVector[Double], Double, Double)] = {
```
Comment (danking) [inline src/main/scala/is/hail/stats/RegressionUtils.scala]: This change is completely independent from the logistic change, it shouldn't be in this PR.

File: src/main/scala/is/hail/methods/LogisticRegressionBurden.scala (hunk 4 above)
Comment (jbloom22) [inline src/main/scala/is/hail/methods/LogisticRegressionBurden.scala]: As opposed to Firth null model which is re-fit per variant / key and implicitly regularized.  I've simplified the language.

File: src/test/scala/is/hail/methods/LinearRegressionBurdenSuite.scala (hunk 6 above)
Comment (jbloom22) [inline src/test/scala/is/hail/methods/LinearRegressionBurdenSuite.scala]: Moved to TestUtils

**Comment by jbloom22** [comment]
@danking back to you
